src/notifier/smtp_notifier.py

```diff
@@ -0,0 +1,73 @@
+# std
+import logging
+from typing import List
+import smtplib
+import email.utils
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+
+
+# project
+from . import Notifier, Event, EventType
+
+
+class SMTPNotifier(Notifier):
+    def __init__(self, title_prefix: str, config: dict):
+        logging.info("Initializing Email notifier.")
+        super().__init__(title_prefix, config)
+        try:
+            self.sender = config["sender"]
+            self.sender_name = config["sender_name"]
+            self.recipient = config["recipient"]
+            self.username_smtp = config["username_smtp"]
+            self.password_smtp = config["password_smtp"]
+            self.host = config["host"]
+            self.port = config["port"]
+
+        except KeyError as key:
+            logging.error(f"Invalid config.yaml. Missing key: {key}")
+
+    def send_events_to_user(self, events: List[Event]) -> bool:
+        errors = False
+        for event in events:
+            if event.type == EventType.USER:
+                priority = ""
+
+                if event.priority == event.priority.HIGH:
+                    priority = "🚨"
+                elif event.priority == event.priority.NORMAL:
+                    priority = "⚠️"
+                subject = f"{priority} {self._title_prefix} {event.service.name} {event.message}"
+                text = ""
+                # Create message container - the correct MIME type is multipart/alternative.
+                msg = MIMEMultipart("alternative")
+                msg["Subject"] = subject
+                msg["From"] = email.utils.formataddr((self.sender_name, self.sender))
+                msg["To"] = self.recipient
+
+                # Record the MIME types of both parts - text/plain and text/html.
+                part1 = MIMEText(text, "plain")
+                part2 = MIMEText(text, "html")
+
+                # Attach parts into message container.
+                # According to RFC 2046, the last part of a multipart message, in this case
+                # the HTML message, is best and preferred.
+                msg.attach(part1)
+                msg.attach(part2)
+
+                # Try to send the message.
+                try:
+                    server = smtplib.SMTP(self.host, self.port)
```

Comment:
I've tested this with credentials for SMTP Relay from Mailjet, and at least with that service, the initialisation of the connection takes **a lot** of time. It takes 185 seconds to complete the test cases.

Did you observe similar delays? Might be better to initialise this connection once in the constructor and keep it alive so that we don't incur time cost of doing that on every notification.

I checked with the current logic, this will block the thread that reads the logs for about 30 seconds but then the parser should be able to catch up, so it might be fine for now.

Reply:
I didn't take that long. AWS SES took ~5 seconds to send the 6 emails. Now trying to re-use the SMTP connection if the connection is still open.